File: 2021/python/day09.py
from collections import namedtuple
import math

sample_input = """
2199943210
3987894921
9856789892
8767896789
9899965678
""".strip()

Grid = namedtuple("Grid", ["grid", "x_max", "y_max"])

with open("input/day09.txt") as f:
    real_input = f.read().strip()


def neighbors(pt, grid):
    [x, y] = pt
    if x > 0:
        yield (x - 1, y)
    if y > 0:
        yield (x, y - 1)
    if x < grid.x_max:
        yield (x + 1, y)
    if y < grid.y_max:
        yield (x, y + 1)


def parse(input):
    out = {}
    lines = input.split("\n")
    for y, line in enumerate(lines):
        for x, d in enumerate(line):
            out[(x, y)] = int(d)
    return Grid(out, len(lines[0]) - 1, len(lines) - 1)


def low_points(inp):
    for pt in inp.grid:
        value = inp.grid[pt]
        ns = [inp.grid[n] for n in neighbors(pt, inp)]
        if all(value < n for n in ns):
            yield pt


def run(input):
    inp = parse(input)
    print(sum(inp.grid[pt] + 1 for pt in low_points(inp)))


def get_basin(pt, inp):
    larger = [
        n for n in neighbors(pt, inp) if inp.grid[n] > inp.grid[pt] and inp.grid[n] != 9
    ]
    return set([pt] + [p for n in larger for p in get_basin(n, inp)])


def run2(input):
    inp = parse(input)
    low = low_points(inp)
    basins = [get_basin(pt, inp) for pt in low_points(inp)]
    basin_sizes = [len(b) for b in sorted(basins, key=len)]
    print(math.prod(basin_sizes[-3:]))


print("=" * 40)
run(sample_input)
run(real_input)

print()
run2(sample_input)
run2(real_input)

File: 2021/ts/day09.ts
import type { Pt } from "./util.ts";
import { getInput, ptKey } from "./util.ts";

const sampleInput = `
2199943210
3987894921
9856789892
8767896789
9899965678
`.trim();

const realInput = (await getInput(import.meta.url)).trim();

function* neighbors(
  x: number,
  y: number,
  maxx: number,
  maxy: number,
): Generator<Pt> {
  if (x > 0) yield [x - 1, y];
  if (y > 0) yield [x, y - 1];
  if (x < maxx) yield [x + 1, y];
  if (y < maxy) yield [x, y + 1];
}

const parse = (input: string) => {
  const lines = input
    .split("\n")
    .map((s) => s.split("").map((n) => parseInt(n, 10)));
  return lines;
};

function getLowPoints(board: number[][]) {
  const maxx = board[0].length - 1;
  const maxy = board.length - 1;

  const lowPoints: Pt[] = [];
  for (let y = 0; y <= maxy; y++) {
    for (let x = 0; x <= maxx; x++) {
      const value = board[y][x];
      const ns = Array.from(neighbors(x, y, maxx, maxy)).map(
        ([x, y]) => board[y][x],
      );
      if (ns.every((v) => v > value)) {
        lowPoints.push([x, y]);
      }
    }
  }
  return lowPoints;
}

function run(input: string) {
  const inp = parse(input);
  console.log(
    getLowPoints(inp)
      .map(([x, y]) => inp[y][x] + 1)
      .reduce((a, b) => a + b),
  );
}

// from a point, fan out to all the higher non-9 points until you run out
function getBasin([x, y]: Pt, board: number[][]): Pt[] {
  const maxx = board[0].length - 1;
  const maxy = board.length - 1;
  const ns = Array.from(neighbors(x, y, maxx, maxy));
  const larger = ns.filter(
    ([nx, ny]) => board[ny][nx] > board[y][x] && board[ny][nx] != 9,
  );
  return [[x, y], ...larger.flatMap((n) => getBasin(n, board))];
}

function run2(input: string) {
  const inp = parse(input);
  const result = getLowPoints(inp)
    .map((pt) => getBasin(pt, inp))
    .map((b) => new Set(b.map(ptKey))) // dedupe
    .sort((a, b) => b.size - a.size)
    .slice(0, 3)
    .map((b) => b.size)
    .reduce((a, b) => a * b);
  console.log(result);
}

console.log("=======================");
run(sampleInput);
run(realInput);

console.log();
run2(sampleInput);
run2(realInput);
